src/features/staking/StakingModal.tsx

```diff
@@ -1,188 +1,157 @@
-    } else {
-      return true;
-    }
-  };
-  const isRenderManageStake = renderManageStake();
+  const isAmountInvalid = Number(stakingAmount) <= 0;
+  const insufficientAstBalance =
+    txType === TxType.STAKE && Number(stakingAmount) > Number(astBalance);
+  const insufficientSastBalance =
+    txType === TxType.UNSTAKE && stakingAmount > unstakableSastBalance;
+
+  const isStakeButtonDisabled =
+    isAmountInvalid || insufficientAstBalance || insufficientSastBalance;
+
+  const successText = (
+    <AmountStakedText
+      stakingAmount={stakingAmount}
+      txStatus={txStatus}
+      dataApproveAst={dataApproveAst}
+      dataStakeAst={dataStakeAst}
+      dataUnstakeSast={dataUnstakeSast}
+    />
+  );
+  const actionDescription = transactionTrackerMessages({
+    txStatus,
+    dataApproveAst,
+    dataStakeAst,
+    dataUnstakeSast,
+  });
 
-  const isRenderApproveSuccess =
-    statusStake === "success" || statusUnstake === "success";
+  const actionButtons = actionButtonsObject({
+    resetApproveAst,
+    resetStakeAst,
+    resetUnstakeSast,
+  });
+
+  const modalLoadingStateHeadlines = modalTxLoadingStateHeadlines({
+    txStatus,
+    dataApproveAst,
+    dataStakeAst,
+    dataUnstakeSast,
+  });
 
-  const handleCloseModal = () => {
-    stakingModalRef.current && stakingModalRef.current.close();
-    setValue("stakingAmount", 0);
+  const actionButtonLogic = () => {
+    if (dataApproveAst) {
+      return actionButtons.approve;
+    } else if (dataStakeAst) {
+      return actionButtons.stake;
+    } else if (dataUnstakeSast) {
+      return actionButtons.unstake;
+    } else {
+      return undefined;
+    }
   };
 
-  // TODO: replace this with our `Modal` component so it inherits the close
-  // to click etc.
   return (
-    <dialog
-      className={twJoin(
-        "bg-gray-900 p-6 text-white",
-        "w-full max-w-none xs:max-w-[360px]",
-        "border border-gray-800 rounded-none xs:rounded-lg",
-        "backdrop:bg-gray-950 backdrop:bg-opacity-[85%] backdrop:backdrop-blur-[2px]",
-      )}
-      ref={stakingModalRef}
+    <Modal
+      className="w-full max-w-none xs:max-w-[360px] text-white"
+      modalHeadline={modalLoadingStateHeadlines}
+      onCloseRequest={() => setShowStakingModal(false)}
     >
-      <div className="flex justify-between items-center mb-7 mt-1">
-        <h2 className="font-semibold text-xl">{headline}</h2>
-        <div className="hover:cursor-pointer" onClick={handleCloseModal}>
-          <MdClose className="text-gray-500" size={26} />
-        </div>
-      </div>
-
-      {isRenderManageStake ? (
-        <ManageStake
-          formReturn={formReturn}
-          stakeOrUnstake={stakeOrUnstake}
-          setStakeOrUnstake={setStakeOrUnstake}
-          statusApprove={statusApprove}
-          statusStake={statusStake}
-          statusUnstake={statusUnstake}
+      {transactionHashes ? (
+        <TransactionTracker
+          actionDescription={actionDescription}
```

Comment:
Similar to `successText`, my feeling is that `TransacitonTracker` should be responsible for conditionally rendering the correct text based on the status of the transaction.

As such, you should simply pass `"stake 30 AST" | "unstake 50 AST" | "approve 30 AST to stake"` (or whatever), then have `TransactionTracker` pull the status and add the rest of the text, i.e.

```typescript
if (status === "loading") {
  text = `Your transaction to ${actionDescription} is processing. Please wait a moment.`
}
```

Having the transaction tracker handle all of the transaction tracking logic reduces the amount of boilerplate needed if used elsewhere.